// jcv-ts-utils/dist/commons/commons.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadImage = exports.createDebounce = exports.notInitializedPromise = exports.notInitializedFn = exports.notImplementFn = void 0;
function notImplementFn() {
    throw new Error("Not Implemented function");
}
exports.notImplementFn = notImplementFn;
function notInitializedFn() {
    throw new Error("Not initialized function");
}
exports.notInitializedFn = notInitializedFn;
async function notInitializedPromise() {
    throw new Error("Not initialized function");
}
exports.notInitializedPromise = notInitializedPromise;
/**
 * Instancie un debounce qui permet de se faire rappeler à la fin du timeout
 * si un autre call s'est exécuté alors le timeout est réinitialisé
 * et la fonction précédente sera annulée
 * @param handler
 * @param time
 */
const createDebounce = (handler, time) => {
    let timeoutId;
    return {
        call: (...args) => {
            window.clearTimeout(timeoutId);
            timeoutId = window.setTimeout(() => {
                handler.apply(null, args);
            }, time);
        },
        abort: () => {
            window.clearTimeout(timeoutId);
        },
    };
};
exports.createDebounce = createDebounce;
function loadImage(url) {
    return new Promise((resolve, reject) => {
        const img = new Image();
        img.onload = () => {
            resolve(img);
        };
        img.onerror = reject;
        img.src = url;
    });
}
exports.loadImage = loadImage;
