// jcv-ts-utils/dist/commons/keyframes-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyframesUtils = void 0;
const fadeIn = [{ opacity: 0 }, { opacity: 1 }];
const fadeOut = [{ opacity: 1 }, { opacity: 0 }];
const slideInFrom = (x = "0", y = "0") => {
    return [
        {
            opacity: 0,
            transform: `translate3d(${x},${y},0)`,
            transformOrigin: "top",
        },
        {
            opacity: 1,
            transform: "translate3d(0,0,0)",
            transformOrigin: "top",
        },
    ];
};
const slideOutFrom = (x = "0", y = "0") => {
    return [
        {
            opacity: 1,
            pointerEvents: "none",
            transform: "translate3d(0,0,0)",
            transformOrigin: "top",
        },
        {
            opacity: 0,
            pointerEvents: "none",
            transform: `translate3d(${x},${y},0)`,
            transformOrigin: "top",
        },
    ];
};
const increaseHeight = (maxHeight) => {
    return [
        {
            maxHeight: "0px",
            overflow: "hidden",
        },
        {
            maxHeight,
            overflow: "hidden",
        },
    ];
};
const increaseWidth = (maxWidth) => {
    return [
        {
            maxWidth: "0px",
            overflow: "hidden",
        },
        {
            maxWidth,
            overflow: "hidden",
        },
    ];
};
const decreaseWidth = (maxWidth) => {
    return [
        {
            maxWidth,
            overflow: "hidden",
        },
        {
            maxWidth: "0px",
            overflow: "hidden",
        },
    ];
};
const decreaseHeight = (maxHeight) => {
    return [
        {
            maxHeight,
            overflow: "hidden",
        },
        {
            maxHeight: "0px",
            overflow: "hidden",
        },
    ];
};
exports.KeyframesUtils = {
    fadeIn,
    fadeOut,
    increaseWidth,
    increaseHeight,
    decreaseHeight,
    decreaseWidth,
    slideInFrom,
    slideOutFrom,
};
